crates/storage/nippy-jar/src/writer.rs

```diff
@@ -0,0 +1,401 @@
+use crate::{compression::Compression, ColumnResult, NippyJar, NippyJarError, NippyJarHeader};
+use std::{
+    cmp::Ordering,
+    fs::{File, OpenOptions},
+    io::{Read, Seek, SeekFrom, Write},
+    path::Path,
+};
+
+/// Size of one offset in bytes.
+const OFFSET_SIZE_BYTES: u64 = 8;
+
+/// Writer of [`NippyJar`]. Handles table data and offsets only.
+///
+/// Table data is written directly to disk, while offsets and configuration need to be flushed by
+/// calling `commit()`.
+///
+/// ## Offset file layout
+/// 1. First byte – size of one offset in bytes
+/// 2. N offsets, each "size of one offset" bytes
+/// 2. Last "size of one offset" bytes – total size of the data file (also, the next offset where
+///    the data should be written next)
```

Comment:
suggestion on how to avoid the "size of one offset" things in the docs

```suggestion
/// The first byte is the size of a single offset in bytes, `m`.
/// Then, the file contains `n` entries, each with a size of `m`. Each entry represents an offset, except for the last entry, which represents both the total size of the data file, as well as the next offset to write new data to.
```